pingcap/pd
# server/schedule: refine metrics about store limit

### What problem does this PR solve? 
refine metrics about store limit: add storeLimitCostCounter and storeLimitAvailableGauge to monitor the status of store limiter.

### What is changed and how it works?
refine metrics about store limit
1. add storeLimitCostCounter to monitor the step cost
2. add storeLimitAvailableGauge to monitor the available status of storeLimit, and it is collected in a background job.
3. add storeLimitRateGauge to show the store limit setting in pd.

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)
1. run a tidb cluste with 4 tikv instances
2. prepare some data using sysbench like : `sysbench --mysql-host=127.0.0.1 --mysql-port=4000 --mysql-user=root --table-size=1000000 oltp_insert prepare`
3. run `pd-ctl -u 0.0.0.0:32774 store limit all 1  region-add`  and `pd-ctl -u 0.0.0.0:32774 store limit all 1  region-remove ` to set the store limit
4. run `pd-ctl -u 0.0.0.0:32774 store delete 3057` to remove one store
5. run `curl 0.0.0.0:32774/metrics | grep store_limit` to get the metrics. The metrics should be like:
```
# HELP pd_schedule_store_limit_available available limit rate of store.
# TYPE pd_schedule_store_limit_available gauge
pd_schedule_store_limit_available{limit_type="region-add",store="1"} 1
pd_schedule_store_limit_available{limit_type="region-add",store="3057"} 1
pd_schedule_store_limit_available{limit_type="region-add",store="4"} 1
pd_schedule_store_limit_available{limit_type="region-add",store="5"} 1
pd_schedule_store_limit_available{limit_type="region-remove",store="1"} 1
pd_schedule_store_limit_available{limit_type="region-remove",store="3057"} 100000
pd_schedule_store_limit_available{limit_type="region-remove",store="4"} 1
pd_schedule_store_limit_available{limit_type="region-remove",store="5"} 1
# HELP pd_schedule_store_limit_cost limit rate cost of store.
# TYPE pd_schedule_store_limit_cost counter
pd_schedule_store_limit_cost{limit_type="region-add",store="1"} 1
pd_schedule_store_limit_cost{limit_type="region-add",store="4"} 2
pd_schedule_store_limit_cost{limit_type="region-remove",store="3057"} 3
# HELP pd_schedule_store_limit_rate the limit rate of store.
# TYPE pd_schedule_store_limit_rate gauge
pd_schedule_store_limit_rate{limit_type="region-add",store="1"} 1.000000016666667
pd_schedule_store_limit_rate{limit_type="region-add",store="3057"} 6e+06
pd_schedule_store_limit_rate{limit_type="region-add",store="4"} 1.000000016666667
pd_schedule_store_limit_rate{limit_type="region-add",store="5"} 1.000000016666667
pd_schedule_store_limit_rate{limit_type="region-remove",store="1"} 1.000000016666667
pd_schedule_store_limit_rate{limit_type="region-remove",store="3057"} 6e+06
pd_schedule_store_limit_rate{limit_type="region-remove",store="4"} 1.000000016666667
pd_schedule_store_limit_rate{limit_type="region-remove",store="5"} 1.000000016666667
```
 
Side effects

 - Breaking backward compatibility
The metrics pd_schedule_store_limit is removed.
